### ISSUE TITLE
Problem in unused parameters and momentum based optimizer

### ISSUE THREAD

**Issue by t-ae** [tensorflow/swift-apis#421]
```swift
import TensorFlow

struct Model: Layer {
    var weight1 = Tensor<Float>(randomNormal: [3])
    var weight2 = Tensor<Float>(randomNormal: [3])
    
    @noDerivative
    var useWeight2: Bool = true
    
    @differentiable
    func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
        var x = input * weight1
        if useWeight2 {
            x = x * weight2
        }
        return x.sum() - 1
    }
}

var model = Model()
var opt = Adam(for: model)

// Update weight1 and weight2
for _ in 0..<3 {
    let tensor = Tensor<Float>(randomNormal: [3])
    let dmodel = model.gradient { model in
        model(tensor)
    }
    opt.update(&model.allDifferentiableVariables, along: dmodel)
}

// Wanna update weight1 only
print("disable weight2")
model.useWeight2 = false
for _ in 0..<3 {
    let tensor = Tensor<Float>(randomNormal: [3])
    let dmodel = model.gradient { model in
        model(tensor)
    }
    opt.update(&model.allDifferentiableVariables, along: dmodel)
    
    // Weight2 is changing!
    print("weight2:", model.weight2)
}
```

In the last `for` block, I expect `model.weight2` not to change.
`dmodel.weight2` is `0.0` here. But `weight2` is updated with momentum of `opt`.
I think it's undesirable bahavior.

**Comment by Shashi456** [comment]
I'm not sure if weight freezing is a functionality that's been added yet.

**Comment by t-ae** [comment]
IMO freezing and unused are different.
Freezing means values won't change even if they are used.
Unused parameters are not in computational graph so it shouldn't be changed.

**Comment by Shashi456** [comment]
@t-ae what happens when you initially set the bool to be false, does that work? That'd maybe explain if the graph is taking in conditions at all.

**Comment by t-ae** [comment]
> what happens when you initially set the bool to be false, does that work?

Yes it works, and `weight2` doesn't change. But it is because momentum is initialized with 0.

I tried similar code in TensorFlow 2.0 and PyTorch.
In PyTorch I got same behavior (weight2 changes even if it's unused).

In TensorFlow the gradient of `weight2` was `None` so its value and optimizer's momentum don't change.
But with `None` gradient, `Adam.apply_gradients` emits warning message.
Probably I'm doing something wrong (I'm not used to TF).

Now I got unconfident it should be fixed or not.
I'll close this issue.

TF code I tried:

```python
import tensorflow as tf

class Model(tf.keras.layers.Layer):
    def __init__(self):
        super().__init__()
        self.weight1 = tf.Variable(tf.random.normal([3]))
        self.weight2 = tf.Variable(tf.random.normal([3]))

        self.use_weight2 = True

    def call(self, input):
        x = input * self.weight1
        if self.use_weight2:
            x *= self.weight2
        return tf.math.reduce_sum(x) - 1

model = Model()
opt = tf.keras.optimizers.Adam()
print(model.trainable_variables)

for i in range(3):
    tensor = tf.random.normal([3])
    with tf.GradientTape() as tape:
        output = model(tensor)
    grads = tape.gradient(output, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))

model.use_weight2 = False
for i in range(3, 6):
    tensor = tf.random.normal([3])
    with tf.GradientTape() as tape:
        output = model(tensor)
    grads = tape.gradient(output, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))

    print(model.weight2)
```
